File: Data Structures & Algorithms/path-sum/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
    private boolean dfs(TreeNode node, int targetSum, int currentSum) {
        if (node == null) return false;
        else {
            currentSum += node.val;
            
            if (node.left == null && node.right == null) return targetSum == currentSum;
            else return dfs(node.left, targetSum, currentSum) || dfs(node.right, targetSum, currentSum);
        }
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        num_islands = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    num_islands += 1
                    
                    def dfs(x: int, y:int):
                        if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[0]) or grid[x][y] != '1': return
                        else:
                            grid[x][y] = '?'

                            dfs(x, y + 1)
                            dfs(x, y - 1)
                            dfs(x + 1, y)
                            dfs(x - 1, y)
                            return
                    
                    dfs(i, j)

        return num_islands
